src/ZODB/Connection.py

```diff
@@ -1083,6 +1042,7 @@ def _resetCache(self):
             self._reader._cache = cache
 
     def _release_resources(self):
+        assert not self.opened
```

Comment:
This code has lots of asserts.  I'm ambivalent about them myself, but again, they're used a lot in ZODB, so I played along. <shrug>

I think they're a terrible idea in test code, unless you're using py.test.